engine/evd.ml

```diff
@@ -584,6 +584,16 @@ let undefined_map d = d.undf_evars
 
 let drop_all_defined d = { d with defn_evars = EvMap.empty }
 
+let reset_non_goal_defined d =
+  let f evk info =
+    match snd info.evar_source with
+    (* These are the two evar kinds used for existing goals *)
+    (* see Proofview.mark_in_evm *)
+    | Evar_kinds.GoalEvar | Evar_kinds.VarInstance _ -> info
+    | _ -> { info with evar_body = Evar_empty } in
+  let defn_evars = EvMap.Smart.mapi f d.defn_evars in
```

Comment:
This ends up with empty bodies in defn_evars, is that really ok?

Reply:
That's the point, yes.

Reply:
I mean shouldn't you move them back to undf_evars?

Reply:
Yes, doing so breaks quite a few invariants of the evarmap structure.

Reply:
You're right. In practice, it worked and I did not investigate further because `existential_value` is "permissive" about it, but the undefined evars should be moved to the right map.